# Lab | Data Cleaning and Formatting

In this lab, we will be working with the customer data from an insurance company, which can be found in the CSV file located at the following link: https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv


# Challenge 1: Data Cleaning and Formatting

## Exercise 1: Cleaning Column Names

To ensure consistency and ease of use, standardize the column names of the dataframe. Start by taking a first look at the dataframe and identifying any column names that need to be modified. Use appropriate naming conventions and make sure that column names are descriptive and informative.

*Hint*:
- *Column names should be in lower case*
- *White spaces in column names should be replaced by `_`*
- *`st` could be replaced for `state`*

In [1]:
import pandas as pd

# Load the CSV file into a pandas DataFrame
url = "https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv"
df = pd.read_csv(url)

# Clean and format the column names
df.columns = df.columns.str.lower().str.replace(' ', '_').str.replace('st', 'state').str.replace('custateomer', 'customer')

# Verify the changes
print(df.head())
print(df.columns)

  customer       state gender             education customer_lifetime_value  \
0  RB50392  Washington    NaN                Master                     NaN   
1  QZ44356     Arizona      F              Bachelor              697953.59%   
2  AI49188      Nevada      F              Bachelor             1288743.17%   
3  WW63253  California      M              Bachelor              764586.18%   
4  GA49547  Washington      M  High School or Below              536307.65%   

    income  monthly_premium_auto number_of_open_complaints     policy_type  \
0      0.0                1000.0                    1/0/00   Personal Auto   
1      0.0                  94.0                    1/0/00   Personal Auto   
2  48767.0                 108.0                    1/0/00   Personal Auto   
3      0.0                 106.0                    1/0/00  Corporate Auto   
4  36357.0                  68.0                    1/0/00   Personal Auto   

   vehicle_class  total_claim_amount  
0  Four-Door Car 

## Exercise 2: Cleaning invalid Values

The dataset contains columns with inconsistent and incorrect values that could affect the accuracy of our analysis. Therefore, we need to clean these columns to ensure that they only contain valid data.

Note that this exercise will focus only on cleaning inconsistent values and will not involve handling null values (NaN or None).

*Hint*:
- *Gender column contains various inconsistent values such as "F", "M", "Femal", "Male", "female", which need to be standardized, for example, to "M" and "F".*
- *State abbreviations be can replaced with its full name, for example "AZ": "Arizona", "Cali": "California", "WA": "Washington"*
- *In education, "Bachelors" could be replaced by "Bachelor"*
- *In Customer Lifetime Value, delete the `%` character*
- *In vehicle class, "Sports Car", "Luxury SUV" and "Luxury Car" could be replaced by "Luxury"*

In [2]:
# Standardize the gender column
df['gender'] = df['gender'].replace({
    'Femal': 'F', 'female': 'F', 'Female': 'F',
    'Male': 'M', 'male': 'M'
})

# Replace state abbreviations with full names
df['state'] = df['state'].replace({
    'AZ': 'Arizona', 'Cali': 'California', 'WA': 'Washington'
})

# Standardize the education column
df['education'] = df['education'].replace('Bachelors', 'Bachelor')

# Remove the percentage sign from 'custateomer_lifetime_value' and convert to numeric
df['customer_lifetime_value'] = df['customer_lifetime_value'].str.rstrip('%').astype('float') / 100

# Standardize the vehicle class column
df['vehicle_class'] = df['vehicle_class'].replace({
    'Sports Car': 'Luxury', 'Luxury SUV': 'Luxury', 'Luxury Car': 'Luxury'
})

# Verify the changes
print(df.head())
print(df.dtypes)

  customer       state gender             education  customer_lifetime_value  \
0  RB50392  Washington    NaN                Master                      NaN   
1  QZ44356     Arizona      F              Bachelor                6979.5359   
2  AI49188      Nevada      F              Bachelor               12887.4317   
3  WW63253  California      M              Bachelor                7645.8618   
4  GA49547  Washington      M  High School or Below                5363.0765   

    income  monthly_premium_auto number_of_open_complaints     policy_type  \
0      0.0                1000.0                    1/0/00   Personal Auto   
1      0.0                  94.0                    1/0/00   Personal Auto   
2  48767.0                 108.0                    1/0/00   Personal Auto   
3      0.0                 106.0                    1/0/00  Corporate Auto   
4  36357.0                  68.0                    1/0/00   Personal Auto   

   vehicle_class  total_claim_amount  
0  Four-Doo

## Exercise 3: Formatting data types

The data types of many columns in the dataset appear to be incorrect. This could impact the accuracy of our analysis. To ensure accurate analysis, we need to correct the data types of these columns. Please update the data types of the columns as appropriate.

It is important to note that this exercise does not involve handling null values (NaN or None).

*Hint*:
- *Customer lifetime value should be numeric*
- *Number of open complaints has an incorrect format. Look at the different values it takes with `unique()` and take the middle value. As an example, 1/5/00 should be 5. Number of open complaints is a string - remember you can use `split()` to deal with it and take the number you need. Finally, since it should be numeric, cast the column to be in its proper type.*

In [3]:
# Clean the 'number_of_open_complaints' column
df['number_of_open_complaints'] = df['number_of_open_complaints'].apply(lambda x: x.split('/')[1] if isinstance(x, str) else x)
df['number_of_open_complaints'] = pd.to_numeric(df['number_of_open_complaints'], errors='coerce')

# Verify the changes
print(df.head())
print(df.dtypes)

  customer       state gender             education  customer_lifetime_value  \
0  RB50392  Washington    NaN                Master                      NaN   
1  QZ44356     Arizona      F              Bachelor                6979.5359   
2  AI49188      Nevada      F              Bachelor               12887.4317   
3  WW63253  California      M              Bachelor                7645.8618   
4  GA49547  Washington      M  High School or Below                5363.0765   

    income  monthly_premium_auto  number_of_open_complaints     policy_type  \
0      0.0                1000.0                        0.0   Personal Auto   
1      0.0                  94.0                        0.0   Personal Auto   
2  48767.0                 108.0                        0.0   Personal Auto   
3      0.0                 106.0                        0.0  Corporate Auto   
4  36357.0                  68.0                        0.0   Personal Auto   

   vehicle_class  total_claim_amount  
0  Fo

## Exercise 4: Dealing with Null values

Identify any columns with null or missing values. Identify how many null values each column has. You can use the `isnull()` function in pandas to find columns with null values.

Decide on a strategy for handling the null values. There are several options, including:

- Drop the rows or columns with null values
- Fill the null values with a specific value (such as the column mean or median for numerical variables, and mode for categorical variables)
- Fill the null values with the previous or next value in the column
- Fill the null values based on a more complex algorithm or model (note: we haven't covered this yet)

Implement your chosen strategy to handle the null values. You can use the `fillna()` function in pandas to fill null values or `dropna()` function to drop null values.

Verify that your strategy has successfully handled the null values. You can use the `isnull()` function again to check if there are still null values in the dataset.

Remember to document your process and explain your reasoning for choosing a particular strategy for handling null values.

After formatting data types, as a last step, convert all the numeric variables to integers.

In [4]:
# Identify columns with null values
null_counts = df.isnull().sum()
print("Null values in each column:\n", null_counts)

# Fill null values with column mean for numerical variables and mode for categorical variables
for column in df.columns:
    if df[column].dtype == 'object':
        df[column] = df[column].fillna(df[column].mode()[0])
    else:
        df[column] = df[column].fillna(df[column].mean())

# Verify that null values have been handled
null_counts_after = df.isnull().sum()
print("Null values in each column after handling:\n", null_counts_after)

# Convert all numeric variables to integers
numeric_columns = df.select_dtypes(include=['float64']).columns
df[numeric_columns] = df[numeric_columns].astype('int')

# Verify the changes
print(df.head())
print(df.dtypes)

Null values in each column:
 customer                     2937
state                        2937
gender                       3054
education                    2937
customer_lifetime_value      2940
income                       2937
monthly_premium_auto         2937
number_of_open_complaints    2937
policy_type                  2937
vehicle_class                2937
total_claim_amount           2937
dtype: int64
Null values in each column after handling:
 customer                     0
state                        0
gender                       0
education                    0
customer_lifetime_value      0
income                       0
monthly_premium_auto         0
number_of_open_complaints    0
policy_type                  0
vehicle_class                0
total_claim_amount           0
dtype: int64
  customer       state gender             education  customer_lifetime_value  \
0  RB50392  Washington      F                Master                     7936   
1  QZ44356     Arizona    

## Exercise 5: Dealing with duplicates

Use the `.duplicated()` method to identify any duplicate rows in the dataframe.

Decide on a strategy for handling the duplicates. Options include:
- Dropping all duplicate rows
- Keeping only the first occurrence of each duplicated row
- Keeping only the last occurrence of each duplicated row
- Dropping duplicates based on a subset of columns
- Dropping duplicates based on a specific column

Implement your chosen strategy using the `drop_duplicates()` function.

Verify that your strategy has successfully handled the duplicates by checking for duplicates again using `.duplicated()`.

Remember to document your process and explain your reasoning for choosing a particular strategy for handling duplicates.

Save the cleaned dataset to a new CSV file.

*Hint*: *after dropping duplicates, reset the index to ensure consistency*.

In [5]:
# Identify duplicate rows
duplicates = df.duplicated()
print("Number of duplicate rows:", duplicates.sum())

# Drop all duplicate rows
df = df.drop_duplicates()

# Reset the index
df = df.reset_index(drop=True)

# Verify that duplicates have been handled
duplicates_after = df.duplicated()
print("Number of duplicate rows after handling:", duplicates_after.sum())

# Save the cleaned dataset to a new CSV file
df.to_csv('cleaned_data.csv', index=False)

# Verify the changes
print(df.head())
print(df.dtypes)

Number of duplicate rows: 2936
Number of duplicate rows after handling: 0
  customer       state gender             education  customer_lifetime_value  \
0  RB50392  Washington      F                Master                     7936   
1  QZ44356     Arizona      F              Bachelor                     6979   
2  AI49188      Nevada      F              Bachelor                    12887   
3  WW63253  California      M              Bachelor                     7645   
4  GA49547  Washington      M  High School or Below                     5363   

   income  monthly_premium_auto  number_of_open_complaints     policy_type  \
0       0                  1000                          0   Personal Auto   
1       0                    94                          0   Personal Auto   
2   48767                   108                          0   Personal Auto   
3       0                   106                          0  Corporate Auto   
4   36357                    68                        

# Bonus: Challenge 2: creating functions on a separate `py` file

Put all the data cleaning and formatting steps into functions, and create a main function that performs all the cleaning and formatting.

Write these functions in separate .py file(s). By putting these steps into functions, we can make the code more modular and easier to maintain.

*Hint: autoreload module is a utility module in Python that allows you to automatically reload modules in the current session when changes are made to the source code. This can be useful in situations where you are actively developing code and want to see the effects of changes you make without having to constantly restart the Python interpreter or Jupyter Notebook kernel.*

In [10]:
%load_ext autoreload
%autoreload 2

import data_cleaning

data_cleaning.main()

The autoreload extension is already loaded. To reload it, use:
  %reload_ext autoreload
  customer       state   gender             education  \
0  RB50392  Washington  Unknown                Master   
1  QZ44356     Arizona        F              Bachelor   
2  AI49188      Nevada        F              Bachelor   
3  WW63253  California        M              Bachelor   
4  GA49547  Washington        M  High School or Below   

   customer_lifetime_value   income  monthly_premium_auto  \
0              7936.902593      0.0                1000.0   
1              6979.535900      0.0                  94.0   
2             12887.431700  48767.0                 108.0   
3              7645.861800      0.0                 106.0   
4              5363.076500  36357.0                  68.0   

   number_of_open_complaints     policy_type  vehicle_class  \
0                        0.0   Personal Auto  Four-Door Car   
1                        0.0   Personal Auto  Four-Door Car   
2            

# Bonus: Challenge 3: Analyzing Clean and Formated Data

You have been tasked with analyzing the data to identify potential areas for improving customer retention and profitability. Your goal is to identify customers with a high policy claim amount and a low customer lifetime value.

In the Pandas Lab, we only looked at high policy claim amounts because we couldn't look into low customer lifetime values. If we had tried to work with that column, we wouldn't have been able to because customer lifetime value wasn't clean and in its proper format. So after cleaning and formatting the data, let's get some more interesting insights!

Instructions:

- Review the statistics again for total claim amount and customer lifetime value to gain an understanding of the data.
- To identify potential areas for improving customer retention and profitability, we want to focus on customers with a high policy claim amount and a low customer lifetime value. Consider customers with a high policy claim amount to be those in the top 25% of the total claim amount, and clients with a low customer lifetime value to be those in the bottom 25% of the customer lifetime value. Create a pandas DataFrame object that contains information about customers with a policy claim amount greater than the 75th percentile and a customer lifetime value in the bottom 25th percentile.
- Use DataFrame methods to calculate summary statistics about the high policy claim amount and low customer lifetime value data. To do so, select both columns of the dataframe simultaneously and pass it to the `.describe()` method. This will give you descriptive statistics, such as mean, median, standard deviation, minimum and maximum values for both columns at the same time, allowing you to compare and analyze their characteristics.

In [11]:
# Review statistics for total claim amount and customer lifetime value
print("Statistics for Total Claim Amount and Customer Lifetime Value:")
print(df[['total_claim_amount', 'customer_lifetime_value']].describe())

# Identify the 75th percentile for total claim amount and the 25th percentile for customer lifetime value
high_claim_threshold = df['total_claim_amount'].quantile(0.75)
low_clv_threshold = df['customer_lifetime_value'].quantile(0.25)

# Create a DataFrame for customers with high policy claim amounts and low customer lifetime values
target_customers = df[(df['total_claim_amount'] > high_claim_threshold) & (df['customer_lifetime_value'] < low_clv_threshold)]

# Calculate summary statistics for the target customers
print("\nSummary Statistics for High Policy Claim Amount and Low Customer Lifetime Value:")
print(target_customers[['total_claim_amount', 'customer_lifetime_value']].describe())

# Display the first few rows of the target customers DataFrame
print("\nTarget Customers DataFrame:")
print(target_customers.head())

Statistics for Total Claim Amount and Customer Lifetime Value:
       total_claim_amount  customer_lifetime_value
count         1072.000000              1072.000000
mean           404.527985              7936.402985
std            292.905011              6422.757351
min              0.000000              2004.000000
25%            201.750000              4053.500000
50%            354.500000              5888.000000
75%            532.000000              8937.500000
max           2893.000000             58166.000000

Summary Statistics for High Policy Claim Amount and Low Customer Lifetime Value:
       total_claim_amount  customer_lifetime_value
count           40.000000                40.000000
mean           723.300000              3415.250000
std            149.216449               569.199291
min            537.000000              2287.000000
25%            632.750000              3042.000000
50%            685.000000              3659.500000
75%            799.000000              

### Insights from the Analysis

1. **Statistics for Total Claim Amount and Customer Lifetime Value:**
   - The average total claim amount is approximately $404.53, with a standard deviation of $292.91.
   - The average customer lifetime value is approximately $7936.40, with a standard deviation of $6422.76.
   - The 75th percentile for total claim amount is $532.00, and the 25th percentile for customer lifetime value is $4053.50.

2. **Summary Statistics for High Policy Claim Amount and Low Customer Lifetime Value:**
   - The average total claim amount for the target customers is approximately $723.30, which is significantly higher than the overall average.
   - The average customer lifetime value for the target customers is approximately $3415.25, which is significantly lower than the overall average.
   - The standard deviation for total claim amount is $149.22, indicating some variability in the claim amounts among these customers.
   - The standard deviation for customer lifetime value is $569.20, indicating less variability in the lifetime values among these customers.

3. **Target Customers DataFrame:**
   - The target customers have a high total claim amount (greater than $532.00) and a low customer lifetime value (less than $4053.50).
   - These customers are spread across different states and have varying levels of education.
   - The income levels of these customers vary, with some having zero income.
   - The number of open complaints for these customers is generally low.
   - The vehicle classes for these customers include SUVs, Four-Door Cars, and Two-Door Cars.

### Potential Areas for Improving Customer Retention and Profitability

1. **Focus on High Claim Customers with Low Lifetime Value:**
   - These customers are costly due to their high claim amounts but are not very valuable in terms of lifetime value.
   - Consider strategies to either reduce their claim amounts or increase their lifetime value.

2. **Customer Education and Engagement:**
   - Engage with these customers to understand their needs and provide better services that could increase their lifetime value.
   - Offer educational resources or incentives to encourage safer driving habits, which could reduce claim amounts.

3. **Targeted Marketing and Offers:**
   - Develop targeted marketing campaigns to offer personalized products or services that could increase the lifetime value of these customers.
   - Consider loyalty programs or discounts for customers who maintain low claim amounts over time.

4. **Review and Adjust Policies:**
   - Review the insurance policies of these customers to identify any patterns or common factors that contribute to high claim amounts.
   - Adjust policies or premiums accordingly to better manage risk and profitability.

By focusing on these areas, you can work towards improving customer retention and profitability, especially for customers who currently have high claim amounts and low lifetime values.